### ISSUE TITLE
[APL][Random] 'ipc timed out for 0x30200000' after S3 without playback/capture. (5%)

### ISSUE THREAD

**Issue by stevyan** [thesofproject/linux#876]
**Summary:**
 'ipc timed out for 0x30200000' after S3 without playback/capture.
_Note_: Following ipc message id occurred for different attempt times: 0x30200000, 0x30030000, 0x50010000, 0x30010000 ...

**To reproduce**
1. sudo ./stress-s3.sh 100 0 0

**Actual result**
error: ipc timed out for 0x30200000 size 36

**Reproduce rate: 5%**

**Test recipe**
date: 2019-04-26
sof (master): 6a84b4e9
linux (topic/sof-dev): f7e9d4a
topology: sof-apl-pcm512x.tplg, sof-apl-nocodec.tplg

**dmesg log**
```
[  664.869840] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[  664.870078] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[  664.870082] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[  665.177549] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x30200000 size 36
[  665.177560] sof-audio-pci 0000:00:0e.0: status: fw entered - code 00000005
[  665.177603] sof-audio-pci 0000:00:0e.0: error: unexpected fault 0x00000000 trace 0x00004000
[  665.177607] sof-audio-pci 0000:00:0e.0: error: waking up any trace sleepers
[  665.177615] sof-audio-pci 0000:00:0e.0: error: failed to load widget type 31 with ID: 5
[  665.177619] sof-audio-pci 0000:00:0e.0: error: failed to restore pipeline after resume -110
[  665.177642] sof-audio-pci 0000:00:0e.0: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[  665.177945] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
[  665.177969] sof-audio-pci 0000:00:0e.0: FW Poll Status: reg=0x1010303 successful
[  665.177973] sof-audio-pci 0000:00:0e.0: FW Poll Status: reg=0x1000303 successful
[  665.177977] sof-audio-pci 0000:00:0e.0: DSP core(s) enabled? 0 : core_mask 3
[  665.177998] sof-audio-pci 0000:00:0e.0: Debug PCIR: 00000010 at  00000044
```
[ipc_30200000.log](https://github.com/thesofproject/linux/files/3119987/ipc_30200000.log)
[ipc_error-slogger.log](https://github.com/thesofproject/linux/files/3119990/ipc_error-slogger.log)

**Comment by ranj063** [comment]
@stevyan would you be able to check if this PR https://github.com/thesofproject/sof/pull/1343 fixes the problem on APL. I was able to run a 100 iterations stress test with nocodec with this.

**Comment by mengdonglin** [comment]
May relevant to #1254

**Comment by stevyan** [comment]
this error still can be observed after patched the PR thesofproject/sof#1343 on APL nocodec. Iterations target: 500 times, failed at the 213rd.

**Test Recipe**
date: 2019-04-29
sof (master): be1331aa ( + PR thesofproject/sof#1343)
linux (topic/sof-dev): af716732
topology: sof-apl-nocodec.tplg

**dmesg log**
```
[ 1224.449230] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[ 1224.449318] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[ 1224.449326] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30200000: GLB_TPLG_MSG: BUFFER_NEW
[ 1224.752588] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x30200000 size 36
[ 1224.752599] sof-audio-pci 0000:00:0e.0: status: fw entered - code 00000005
[ 1224.752642] sof-audio-pci 0000:00:0e.0: error: unexpected fault 0x00000000 trace 0x00004000
[ 1224.752647] sof-audio-pci 0000:00:0e.0: error: waking up any trace sleepers
[ 1224.752656] sof-audio-pci 0000:00:0e.0: error: failed to load widget type 31 with ID: 31
[ 1224.752661] sof-audio-pci 0000:00:0e.0: error: failed to restore pipeline after resume -110
[ 1224.752688] sof-audio-pci 0000:00:0e.0: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[ 1224.752906] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
[ 1224.752921] sof-audio-pci 0000:00:0e.0: FW Poll Status: reg=0x1010303 successful
[ 1224.752928] sof-audio-pci 0000:00:0e.0: FW Poll Status: reg=0x1000303 successful
[ 1224.752936] sof-audio-pci 0000:00:0e.0: DSP core(s) enabled? 0 : core_mask 3
[ 1224.752950] sof-audio-pci 0000:00:0e.0: Debug PCIR: 00000010 at  00000044
[ 1224.796758] ACPI: Preparing to enter system sleep state S3
[ 1224.797410] PM: Saving platform NVS memory
[ 1224.797418] Disabling non-boot CPUs ...
[ 1224.799315] smpboot: CPU 1 is now offline
[ 1224.801682] smpboot: CPU 2 is now offline
[ 1224.803940] smpboot: CPU 3 is now offline
```

[test_213_fail.log](https://github.com/thesofproject/linux/files/3126208/test_213_fail.log)

**Comment by keyonjie** [comment]
hi @stevyan can you share your ./stress-s3.sh please, I want to duplicate it on my side.

**Comment by keyonjie** [comment]
btw, @stevyan can you paste the sof-logger error log(without -t) also for the error case?

**Comment by stevyan** [comment]
test script attached, e.g.:  #sudo ./stress-s3.sh 500 0 0

[stress-s3.sh](https://github.com/thesofproject/linux/files/3126267/stress-s3.sh.txt)

**Comment by keyonjie** [comment]
both the 2 failed log show we have ipc timeout at runtime resume, so I am curious if we can reproduce it once we disable runtime suspend:
echo on > /sys/bus/pci/devices/0000\:00\:0e.0/power/control

**Comment by keyonjie** [comment]
@stevyan after I disabled runtime suspend, I can't duplicate the issue with 2000 iteration finished.
@ranj063 looks there is some corner cases left in our runtime suspend/resume?

**Comment by ranj063** [comment]
@keyonjie there is aboslutely no difference in sequence between runtime suspend/resume and system suspend/resume in SOF. So I dont know how this test is helpful in identifying the problem

**Comment by keyonjie** [comment]
> @keyonjie there is aboslutely no difference in sequence between runtime suspend/resume and system suspend/resume in SOF. So I dont know how this test is helpful in identifying the problem

I meant doing suspend at SOF runtime-suspended case:
to response to S3 enter command, we wake up it to the active status, and then suspend it.
The issue happens at this waking up.

**Comment by ranj063** [comment]
Yes, but what i meant was that waking up from runtime suspend or waking up from S3 are exactly the same in the SOF driver.

**Comment by Jiangxinx** [comment]
Daily test result:  can still be reproduced on apl_up2_nocodec platform, the rate is 17/100.
By the way, I can't reproduce it on apl_up2_pcm512x platform within 100 times.

***To reproduce***
1. sudo ./stress-s3.sh 100 0 0

***Test Recipe***
sof (master): de5bba49
linux (topic/sof-dev): 1071743
topology: sof-apl-pcm512x.tplg, sof-apl-nocodec.tplg

**Comment by wwittbrx** [comment]
Can you share the config for kernel? I'm getting different error message altogether for this (log attached). This happens always on first iteration of stress-s3.sh

### Recipe

**Kernel**
https://github.com/thesofproject/linux/commit/1071743ab1cf1f3e3653088d36967589c916bc5e

**FW and TPLG**
https://github.com/thesofproject/sof/commit/de5bba49aaf2c0060a279de09580ac4205f550cb
sof-apl-nocodec.tplg

### Reproduction steps

1. sudo ./stress-s3.sh 100 0 0

### Logs and files

[test_1_fail.log](https://github.com/thesofproject/linux/files/3147493/test_1_fail.log)

**Comment by lyakh** [comment]
@wwittbrx could you share your kernel config? I think the kernel configuration, that "should" be used with SOF, or that's used for testing is the one, obtained as described in https://thesofproject.github.io/latest/getting_started/build-guide/build-from-scratch.html#build-linux-kernel

**Comment by wwittbrx** [comment]
This is exactly what I use for config generation. And yet I managed to get different results. My config is attached below.

[conf.config.txt](https://github.com/thesofproject/linux/files/3147738/conf.config.txt)

**Comment by lyakh** [comment]
I plan to look at @wwittbrx 's configuration and why he's seeing a different problem, but first I'm trying to exhaust my own options. While doing my own tests I found out, that the problem for me occurs usually after around 165 suspend-resume loops - sometimes 164. sometimes 167... It happened already at least 3 times to me, so, it isn't likely a coincidence. I would suspect a memory leak, but the firmware is every time loaded fresh, so it cannot be that... A memory leak in the kernel doesn't seem a likely reason either - the Up^2 has 4G of RAM... Just in case this rings a bell?

**Comment by lyakh** [comment]
Update: I modified the driver to fail suspend if runtime-resume fails. That way after suspend fails I can still dump the firmware trace. From it it looks like the firmware does send the last IPC reply, which the host doesn't receive, thus running on a timeout. I'm now increasing the timeout to see, whether the IPC is only taking longer or it goes missing completely.

**Comment by lgirdwood** [comment]
@lyakh hmm, and 165 is not near a power of 2 either.... btw, the "dont suspend if resume fails" sounds like a good PR for debug. Fwiw, DSP and HDA share the same IRQ, maybe the IPC is occurring but HDA/DSP handlers are ignoring it (due to unexpected register values?)

**Comment by lyakh** [comment]
@lgirdwood I think it isn't the host's fault. I'm seeing some *very long* intervals on the DSP for IPC processing:
```
    0      2          IPC             92244.635417     52146.769531 ers/intel/cavs/ipc.c:169 	IPC complete: 0
```
or
```
    0      2       BUFFER             62489.947917     24994.009766 f/src/audio/buffer.c:54     buffer_new()
````
So looks like the DSP is suddenly taking very long to process various random IPC messages. Actually to perform various functions, like creating buffers or enabling a core (this is run with a UP configuration, so that should be a NOP).

**Comment by lyakh** [comment]
@wwittbrx I think it would be good if you could now have a look at this issue too, the ball seems to be in the firmware's court :-)

**Comment by xiulipan** [comment]
@lgirdwood @lyakh 
I think just need a feature more like **do not power off core when error happen**.
@lyakh 
For the timeout here. Do you mean **FW handled the IPC after the Kernel timed out** or **FW handled the IPC but kernel just ignore and timed out**

In first case, please add log to IRQ handler and the IPC task to get the delta. We may still have some race condition in IPC task scheuler.

In second case, please check if the Kernel IRQ hanlder got the IRQ. If get no IRQ check about IPC related mask state. If get the IRQ, sometimes the IRQ thread maybe blocked by previous IRQ. Please check https://github.com/thesofproject/linux/issues/808#issuecomment-482142838 for reference.

**Comment by lyakh** [comment]
@wwittbrx I tried your kernel configuration - I'm getting the same behaviour as with mine, so, that isn't the difference...

**Comment by Jiangxinx** [comment]
This issue can still be reproduced on the daily test,however i can also aplay even though observed timeout error.

***To reproduce***

    1.sudo ./stress-s3.sh 100 0 0
```
---- test 18 of 500 ----
System will suspend after 3 seconds ...
system will resume after 1 seconds ...
rtcwake: wakeup from "mem" using /dev/rtc0 at Wed May  8 01:53:58 2019
Suspend/resume failed, see test_18_fail.log for details
Home directory not accessible: Permission denied
Home directory not accessible: Permission denied
Playing raw data '/dev/zero' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
aplay is OK after 500 times suspend/resume.
== end time: 2019-05-08 09:54:13 ==
ubuntu@cd-apl-up2-pcm512x-02:~/scripts$ cd logs
ubuntu@cd-apl-up2-pcm512x-02:~/scripts/logs$ grep -inr "error" *
test_18_fail.log:196:[  165.268589] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x80010000 size 120
test_18_fail.log:198:[  165.268655] sof-audio-pci 0000:00:0e.0: error: unexpected fault 0x00000000 trace 0x00004000
test_18_fail.log:199:[  165.268661] sof-audio-pci 0000:00:0e.0: error: host status 0x00000000 dsp status 0x00000000 mask 0x00000003
test_18_fail.log:200:[  165.268664] sof-audio-pci 0000:00:0e.0: error: waking up any trace sleepers
test_18_fail.log:201:[  165.268668] sof-audio-pci 0000:00:0e.0: error: failed to set dai config for iDisp2
test_18_fail.log:202:[  165.268671] sof-audio-pci 0000:00:0e.0: error: failed to restore pipeline after resume -110
```
   2.aplay -Dhw:0,0 -r48000 -c2 -fs16_le /dev/zero -vv -i
```
Playing raw data '/dev/zero' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Hardware PCM card 0 'sof-bxt-pcm512x' device 0 subdevice 0
Its setup is:
  stream       : PLAYBACK
  access       : RW_INTERLEAVED
  format       : S16_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 16
  buffer_size  : 16368
  period_size  : 4080
  period_time  : 85000
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 4080
  period_event : 0
  start_threshold  : 16368
  stop_threshold   : 16368
  silence_threshold: 0
  silence_size : 0
  boundary     : 9214364837600034816
  appl_ptr     : 0
  hw_ptr       : 0
#+                                                 | 00%
```

***Test Recipe***
sof (master): 05d4e190
linux (topic/sof-dev): 820e8413
topology: sof-apl-pcm512x.tplg, sof-apl-nocodec.tplg

**Comment by lyakh** [comment]
> @lgirdwood @lyakh
> I think just need a feature more like **do not power off core when error happen**.

Sorry, not sure I understand you correctly. Are you proposing to execute the suspend on the host, but not to shutdown the DSP core, i.e. not to call hda_dsp_core_reset_power_down()? Then the system would suspend, the resume presumably will succeed and the trace will contain the data from before the suspend?

> @lyakh
> For the timeout here. Do you mean **FW handled the IPC after the Kernel timed out** or **FW handled the IPC but kernel just ignore and timed out**

It isn't very easy to exactly identify which step is failing. From what I've seen so far it seems, that the FW always receives the interrupt from the incoming IPC from the host. But then it *looks* like the DSP takes too long before it actually sends out a reply. E.g. a failure case, that I mentioned about on the DSP looks like
```
    0      2          IPC             37073.750000         2.812500 of/src/ipc/handler.c:944    ipc: pipe 1 comp 0 -> new (type 1)
    0      2         HOST             37078.697917         4.947917 sof/src/audio/host.c:350    host_new()
    0      2          DMA             37094.218750        15.520833 intel/cavs/hda-dma.c:822    hda-dmac :5 -> probe
    0      2          DMA             37109.270833        15.052083 rc/sof/src/lib/dma.c:142    dma_get() ID 5 sref = 1 busy channels 0
    0      2          IPC             37321.093750       211.822922 of/src/ipc/handler.c:941    ipc: hdr 0x30010000 tx (84) > rx (28)
    0      2          IPC             37323.958333         2.864583 of/src/ipc/handler.c:944    ipc: pipe 1 comp 1 -> new (type 5)
    0      2       VOLUME             37328.593750         4.635417 f/src/audio/volume.c:161    volume_new()
    0      2          IPC             37495.937500       167.343750 of/src/ipc/handler.c:974    ipc: pipe 1 buffer 2 -> new (0x300 bytes)
    0      2       BUFFER             62489.947917     24994.009766 f/src/audio/buffer.c:54     buffer_new()
```
and on the host it corresponds to
```
[  556.192303] sof-audio-pci 0000:00:0e.0: ipc tx: 0x90010000
[  556.193207] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x90010000
[  556.193220] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30010000
[  556.193462] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x30010000
[  556.193471] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30010000
[  556.193641] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x30010000
[  556.193647] sof-audio-pci 0000:00:0e.0: ipc tx: 0x30200000
[  556.511518] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x30200000 size 36
```
i.e. the DSP receives the last 0x30200000 IPC but it then blocks for a very long time before sending out a reply.

> In first case, please add log to IRQ handler and the IPC task to get the delta. We may still have some race condition in IPC task scheuler.

I've done that, but since then I haven't got any such long interval cases. Instead the last couple of cases were when the host reported receiving a trace position IPC from the DSP, so, that indicates a possible race: an IPC reply seems to be lost when a new IPC message is sent in parallel:

```
[  455.089260] sof-audio-pci 0000:00:0e.0: ipc tx: 0x50010000: GLB_COMP_MSG: SET_VALUE: SET_VALUE
[  455.089353] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x50010000: GLB_COMP_MSG: SET_VALUE: SET_VALUE
[  455.089361] sof-audio-pci 0000:00:0e.0: ipc tx: 0x50010000: GLB_COMP_MSG: SET_VALUE: SET_VALUE
[  455.091088] sof-audio-pci 0000:00:0e.0: ipc rx: 0x90020000: GLB_TRACE_MSG
[  455.091100] sof-audio-pci 0000:00:0e.0: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  455.591070] sof-audio-pci 0000:00:0e.0: ipc rx: 0x90020000: GLB_TRACE_MSG
[  455.591083] sof-audio-pci 0000:00:0e.0: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  455.992984] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x50010000 size 108
```

**Comment by xiulipan** [comment]
@lyath Thanks for the testing. The 0x00032000 timed out looks just like there are some issue inside few to make the handler called after timeout.

For the 0x51000 one, maybe you need to check  if FW do the ipc handler. It seems you also need add some log in kernel IRQ handler.

**Comment by lyakh** [comment]
Here's the best I was able to get today: the kernel view
```
[  467.855920] sof-audio-pci 0000:00:0e.0: DSP core(s) enabled? 1 : core_mask 1
[  467.855926] sof-audio-pci 0000:00:0e.0: ipc tx: 0x40070000: GLB_PM_MSG: CORE_ENABLE
[  468.357539] sof-audio-pci 0000:00:0e.0: ipc rx: 0x90020000: GLB_TRACE_MSG
[  468.357553] sof-audio-pci 0000:00:0e.0: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  468.780402] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x40070000 size 12
```
The FW view:
```
    0      2          IPC             40544.583333        54.166668 ers/intel/cavs/ipc.c:96  	IPC IRQ
    0      2          IPC             40555.208333        10.625000 ers/intel/cavs/ipc.c:148 	IPC task entry
    0      2          IPC             40559.479167         4.270833 of/src/ipc/handler.c:675 	ipc: pm core mask 0x1 -> enable
    0      2          IPC            130670.781250     90111.304688 ers/intel/cavs/ipc.c:173 	IPC complete: 0
```
There are a couple of added trace entries in the FW log. The first one is in `ipc_irq_handler()` and the last one is in `ipc_platform_do_cmd()` which is where the BUSY bit is set, when replying to an IPC message. On the kernel side we also see a received "trace position" message, to which I also added a print in the FW, but it isn't showing up for some reason. And there's again a very long delay on the FW side before raising an IRQ on the host side. So it really looks to me like the FW gets stuck for some long time: either waiting for an interrupt or due to a bug in scheduling.

**Comment by lyakh** [comment]
Here's my best log of the day: on the kernel side
```
[  467.855926] sof-audio-pci 0000:00:0e.0: ipc tx: 0x40070000: GLB_PM_MSG: CORE_ENABLE
[  468.357539] sof-audio-pci 0000:00:0e.0: ipc rx: 0x90020000: GLB_TRACE_MSG
[  468.357553] sof-audio-pci 0000:00:0e.0: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  468.780402] sof-audio-pci 0000:00:0e.0: error: ipc timed out for 0x40070000 size 12
```
on the DSP side
```
    0      2          IPC             40544.583333        54.166668 ers/intel/cavs/ipc.c:96  	IPC IRQ
    0      2          IPC             40555.208333        10.625000 ers/intel/cavs/ipc.c:148 	IPC task entry
    0      2          IPC             40559.479167         4.270833 of/src/ipc/handler.c:675 	ipc: pm core mask 0x1 -> enable
    0      2          IPC            130670.781250     90111.304688 ers/intel/cavs/ipc.c:173 	IPC complete: 0
```
Note received "trace position" messages in the kernel log and a very long delay in the DSP log. Both "IPC task entry" and "IPC complete" messages are printed from `ipc_platform_do_cmd()` which means, that the FW has blocked *inside* that function for that time. The function itself doesn't seem to have obvious possibilities for hanging, which means it was probably interrupted.
As far as I understand trace position messages can only be sent from `do_task_master_core()` after it returns from `wait_for_interrupt()`. But that would only happen after the IPC RX and the scheduled ipc_process_task return? I need to research more to fully understand that flow...

**Comment by lyakh** [comment]
Status update: since I'm often seeing very long delays before the last message in the firmware trace, I began to suspect, that an interrupt is occurring in that time. To verify that I added code to record all interrupts, that might occur in that interval and I dump them with that last log entry. There is always a scheduler and a low-priority task interrupts there, used to schedule the IPC task. Sometimes there are also one or two more interrupts. After multiple runs I've captured such a log, that clearly shows the last IPC with a 50ms delay, which is much longer than typical (under 50us) time offsets for that log entry, but still it should be well within the timeout, that is currently set to 900ms. So it appears to be an instance of (1) unusually long but not too long a delay on the firmware side, and (2) a missing interrupt on the host side with no apparent reason.

**Comment by lyakh** [comment]
Three more observations:
1. Sometimes during an S3 stress test the whole system hangs, requiring a hard reset
2. In about 20-40% of the cases when an IPC timeout occurs, the logger returns an empty trace
3. Even in "normal" cases, when upon an IPC timeout a trace can be retrieved, it is often incomplete. But this might have a logical explanation, related to the IPC timeout itself

**Comment by lyakh** [comment]
Traces seem to be corrupted sometimes:
```
    0      2          IPC             51784.895833         4.218750 of/src/ipc/handler.c:1028   ipc: pipe 23 -> complete
    0      2         PIPE 4.23        51805.885417        20.989584 src/audio/pipeline.c:188    pipeline_complete()
    0      2          IRQ             51809.895833         4.010417 f/src/lib/irqdebug.c:33     IRQ #0: 0x8
    0      2          IRQ             51812.239583         2.343750 f/src/lib/irqdebug.c:33     IRQ #1: 0x3
    0      2          IPC             51814.583333         2.343750 ers/intel/cavs/ipc.c:197    IPC complete: 16777215
````
where the last line should print the return code from `ipc_cmd()` for the pipeline-completion IPC, which can only return 0 or a negative error code.

**Comment by lyakh** [comment]
I'm looking for new ideas. I see the following:
1. The system is preparing to suspend
2. For this the sound card is (runtime-) woken up
3. The driver performs the DSP initialisation, including sending a topology to it
4. Sending the topology involves a strictly serialised repeating "sending of IPC -> waiting for reply" sequence
5. At some point one of IPC replies fails to arrive, which means there is no interrupt and the interrupt registers don't show any bits set, when checked after the timeout expiration (@lgirdwood that's an answer to your question)
6. The DSP log looks either completely or "almost" completely normal: it receives the last IPC and sends a reply.

To me this looks like the hardware is failing to execute the IPC transfer from the DSP to the host. This also seems to be an intermittent failure, because if it is ignored, after it communication with the DSP is possible again. What is the procedure for debugging such cases? I've been told we don't have JTAG access to the DSP, maybe only to the SoC. I'm not sure if we have any GPIOs or other pins available to toggle them from both the DSP and thee host side to check them with a scope. Maybe we could write some data in the mailbox from the DSP side to check it from the host side? E.g. after each received and answered IPC start a periodic task, writing an IPC counter and a repeat counter into the mailbox and then check them on the host side when the timeout expires...

**Comment by lyakh** [comment]
I'm trying to exchange some data over the DSP-to-host mailbox, using e.g.
```
	mailbox_dspbox_write(0x250, &cnt, sizeof(cnt));
```
on the DSP side and
```
	sof_mailbox_read(sdev, 0x250, &cnt, sizeof(cnt));
```
on the host side, but not getting the expected result...

**Comment by xiulipan** [comment]
@lyakh 
Yes, every time you will get DSP reset after a suspend/resume. And we will power off the DSP so everything in memory window is gone. If you would like to have some count, make it in some register. There are few registers that are not using, or you can just use some bits of the register.

**Comment by lyakh** [comment]
@xiulipan I'm not trying to keep the counter across suspends, I'm trying to send it from the DSP to the host. Can I use registers for that too without disturbing the IPC? Which ones?

**Comment by lyakh** [comment]
For the record: following @lgirdwood 's idea to switch over from MSI to "legacy" interrupt mode didn't fix the problem.

**Comment by lyakh** [comment]
update: mailbox is working now.

**Comment by lyakh** [comment]
500 iterations - check. I think I've fixed it, expect a patch soon.

**Comment by wenqingfu** [comment]
please keep the test running while cooking your patch. last year we did 10k iterations to truly prove it's gone :)

**Comment by lyakh** [comment]
fixed by #943

**Comment by xiulipan** [comment]
@lyakh It seems we have a patch for this in FW now https://github.com/thesofproject/sof/pull/1441

**Comment by wenqingfu** [comment]
> fixed by #943

close
